forge-game/src/main/java/forge/game/ability/effects/DiscoverEffect.java

```diff
@@ -0,0 +1,170 @@
+import forge.game.card.CardCollection;
+import forge.game.card.CardLists;
+import forge.game.card.CardZoneTable;
+import forge.game.cost.CostDiscard;
+import forge.game.cost.CostPart;
+import forge.game.cost.CostReveal;
+import forge.game.player.Player;
+import forge.game.player.PlayerController;
+import forge.game.spellability.LandAbility;
+import forge.game.spellability.SpellAbility;
+import forge.game.zone.PlayerZone;
+import forge.game.zone.Zone;
+import forge.game.zone.ZoneType;
+import forge.util.CardTranslation;
+import forge.util.Localizer;
+
+import java.util.List;
+import java.util.Map;
+
+public class DiscoverEffect extends SpellAbilityEffect {
+
+    @Override
+    protected String getStackDescription(SpellAbility sa) {
+        final StringBuilder sb = new StringBuilder();
+        sb.append(sa.getActivatingPlayer()).append(" discovers ");
+        sb.append(sa.getParamOrDefault("Num", "1")).append(".");
+
+        return sb.toString();
+    }
+
+    @Override
+    public void resolve(SpellAbility sa) {
+        final Card host = sa.getHostCard();
+        final Game game = host.getGame();
+        final Player p = sa.getActivatingPlayer();
+
+        // Exile cards from the top of your library until you exile a nonland card with <N> mana value or less.
+        final String val = sa.getParamOrDefault("Num", "1");
+        final int num = AbilityUtils.calculateAmount(host, val, sa);
+
+        if (p == null || !p.isInGame()) return;
+
+        Card found = null;
+        CardCollection exiled = new CardCollection();
+        CardCollection rest = new CardCollection();
+
+        final PlayerZone library = p.getZone(ZoneType.Library);
+
+        for (int i = 0; i < library.size(); i++) {
+            final Card c = library.get(i);
+            exiled.add(c);
+            if (c.isValid("Card.nonLand+cmcLE" + val, sa.getActivatingPlayer(), host, sa)) {
+                found = c;
+                if (sa.hasParam("RememberDiscovered"))
+                    host.addRemembered(c);
+                break;
+            } else {
+                rest.add(c);
+            }
+        }
+
+        if (exiled.size() > 0) {
+            game.getAction().reveal(exiled, p, false);
+        }
+
+        changeZone(exiled, ZoneType.Exile, game, sa);
+
+        // Cast it without paying its mana cost or put it into your hand.
+        if (found != null) {
+            String prompt = Localizer.getInstance().getMessage("lblDiscoverChoice",
+                            CardTranslation.getTranslatedName(found.getName()));
+            final Zone origin = found.getZone();
+            final boolean play = p.getController().chooseBinary(sa, prompt,
+                    PlayerController.BinaryChoiceType.CastOrHand);
+            boolean cancel = false;
+
+            if (play) {
+                // get basic spells (no flashback, etc.)
+                List<SpellAbility> sas = AbilityUtils.getBasicSpellsFromPlayEffect(found, p, CardStateName.Original);
```

Comment:
isn't the third parameter wrong? Doesn't this function only allows two?

i can't find an example for this?

Reply:
Hmm, that would have been copied from PlayEffect... looks like third param is OK, but in this case it's the default anyway so might as well trim it
![image](https://github.com/Card-Forge/forge/assets/103371817/55427df1-d651-465f-ac0e-c5720e6e050f)

Reply:
It was a copy/paste of line 305 in PlayEffect that I must have modified without checking the AbilityUtils options